Rules/CompatibilityRules/UseCompatibleTypes.cs

```diff
@@ -0,0 +1,395 @@
+
+            private readonly List<DiagnosticRecord> _diagnosticAccumulator;
+
+            private readonly UseCompatibleTypes _rule;
+
+            public TypeCompatibilityVisitor(
+                string analyzedFileName,
+                CompatibilityProfileData anyProfile,
+                IReadOnlyList<CompatibilityProfileData> compatibilityTargetProfiles,
+                UseCompatibleTypes rule)
+            {
+                _analyzedFileName = analyzedFileName;
+                _anyProfile = anyProfile;
+                _compatibilityTargets = compatibilityTargetProfiles;
+                _rule = rule;
+                _diagnosticAccumulator = new List<DiagnosticRecord>();
+            }
+
+            public override IEnumerable<DiagnosticRecord> GetDiagnosticRecords()
+            {
+                return _diagnosticAccumulator;
+            }
+
+            public override AstVisitAction VisitTypeExpression(TypeExpressionAst typeExpressionAst)
+            {
+                TryFindTypeIncompatibilities(typeExpressionAst.TypeName);
+                return AstVisitAction.SkipChildren;
+            }
+
+            public override AstVisitAction VisitTypeConstraint(TypeConstraintAst typeConstraintAst)
+            {
+                TryFindTypeIncompatibilities(typeConstraintAst.TypeName);
+                return AstVisitAction.SkipChildren;
+            }
+
+            public override AstVisitAction VisitAttribute(AttributeAst attributeAst)
+            {
+                if (attributeAst.TypeName != null)
+                {
+                    TryFindTypeIncompatibilities(attributeAst.TypeName);
+                }
+
+                return AstVisitAction.SkipChildren;
+            }
+
+            public override AstVisitAction VisitCommand(CommandAst commandAst)
+            {
+                string commandName = commandAst?.GetCommandName();
+                if (commandName == null)
+                {
+                    return AstVisitAction.Continue;
+                }
+
+                if (commandName.Equals("New-Object"))
+                {
+                    if (!TryGetArgument(commandAst.CommandElements, 0, "TypeName", out CommandElementAst typeNameArg))
+                    {
+                        return AstVisitAction.Continue;
+                    }
+
+                    if (!(typeNameArg is StringConstantExpressionAst typeNameStringExp))
+                    {
+                        return AstVisitAction.Continue;
+                    }
+
+                    TryFindTypeIncompatibilities(typeNameStringExp.Value, typeNameStringExp.Extent);
+                    return AstVisitAction.Continue;
+                }
+
+                return AstVisitAction.Continue;
+            }
+
+            private bool TryGetArgument(IReadOnlyList<CommandElementAst> commandElements, int argumentPosition, string argumentParameterName, out CommandElementAst argument)
```

Comment:
This is unfortunately quite complicated, and without the metadata for the command being examined it's technically not correct. But it should work fairly well for most purposes I think